disnake/interactions/base.py

```diff
@@ -679,23 +734,30 @@ async def defer(self, *, ephemeral: bool = False, with_message: bool = False) ->
         if self._responded:
             raise InteractionResponded(self._parent)
 
-        defer_type: int = 0
-        data: Optional[Dict[str, Any]] = None
+        defer_type: Optional[int] = None
+        data: Dict[str, Any] = {}
         parent = self._parent
-        if (
-            parent.type in (InteractionType.application_command, InteractionType.modal_submit)
-            or with_message
-        ):
+        if with_message is MISSING:
+            if ephemeral is not MISSING:
+                with_message = True
+
+        if parent.type in (InteractionType.application_command, InteractionType.modal_submit):
             defer_type = InteractionResponseType.deferred_channel_message.value
-            if ephemeral:
-                data = {"flags": 64}
         elif parent.type is InteractionType.component:
-            defer_type = InteractionResponseType.deferred_message_update.value
+            if with_message:
+                defer_type = InteractionResponseType.deferred_channel_message.value
+            else:
+                defer_type = InteractionResponseType.deferred_message_update.value
+
+        if defer_type == InteractionResponseType.deferred_channel_message.value:
+            data["flags"] = 0
+            if ephemeral is not MISSING:
+                data["flags"] |= MessageFlags.ephemeral.flag
 
         if defer_type:
             adapter = async_context.get()
             await adapter.create_interaction_response(
-                parent.id, parent.token, session=parent._session, type=defer_type, data=data
+                parent.id, parent.token, session=parent._session, type=defer_type, data=data or None
```

Comment:
Have you considered altering `defer_type` from an int (and using .value everywhere) to just the enum item instead? Then converting to .value when eventually passed to the api? It's more flexible, more readable, and is a more intended use of the Enum class-type.

Reply:
https://github.com/DisnakeDev/disnake/pull/474/commits/027af79bd10c3a760103230b8f89431b1dea4d52